src/Juvix/Core/HR/Types.hs

```diff
@@ -1,21 +1,56 @@
 module Juvix.Core.HR.Types where
 
+import qualified Juvix.Core.IR.Types as IR
 import Juvix.Core.Usage
 import Juvix.Library
 
---checkable terms
-data Term primTy primVal
-  = Star Natural --sort i
-  | PrimTy primTy --primitive type
-  | Pi Usage (Term primTy primVal) (Term primTy primVal) --function type
-  | Lam Symbol (Term primTy primVal) --abstraction
-  | Elim (Elim primTy primVal) --elimination
-  deriving (Show, Eq, Generic)
-
---inferable terms
-data Elim primTy primVal
-  = Var Symbol --variable
-  | Prim primVal --primitive constant
-  | App (Elim primTy primVal) (Term primTy primVal) --application
-  | Ann Usage (Term primTy primVal) (Term primTy primVal) --type & usage annotation
-  deriving (Show, Eq, Generic)
+
+data HR
+
+instance IR.TEExt HR where
+  -- FIXME: shouldn't XPi also be a name?
+  type XLam HR primTy primVal = Symbol
+
+  -- (disable these constructors)
+  type XFree HR primTy primVal = Void
+  type XBound HR primTy primVal = Void
+  type ElimX HR primTy primVal = Symbol
+
+
+type Term = IR.Term' HR
+
+pattern Star :: Natural -> Term primTy primVal
```

Comment:
Thanks for writing these. It would be cool if there were a way to *derive* them, do you know if that is possible?

(without Template Haskell, it isn't worth Template Haskell)

Reply:
I think we can do these via, `deriving via`, as they seem to fall directly out of the type itself?

Though I think even that approach may be a bit of a pain.

Reply:
I thought `deriving via` could only produce instances? As far as I know the only way to generate other kinds of declarations is with TH, unfortunately.